src/nanoemoji/write_font.py

```diff
@@ -576,23 +576,42 @@ def _colr_ufo(
     # We want to mutate our view of color glyphs
     color_glyphs = list(color_glyphs)
 
-    # Sort colors so the index into colors == index into CPAL palette.
     # We only store opaque colors in CPAL for COLRv1, as 'alpha' is
     # encoded separately.
-    colors = sorted(
-        set(
-            c if colr_version == 0 else c.opaque()
-            for c in chain.from_iterable(g.colors() for g in color_glyphs)
-            if not c.is_current_color()
-        )
+    colors = set(
+        c if colr_version == 0 else c.opaque()
+        for c in chain.from_iterable(g.colors() for g in color_glyphs)
+        if not c.is_current_color()
     )
+    # Sort colors so the index into colors == index into CPAL palette.
+    # Try to keep the original palette entry index if present.
+    iter_colors_without_index = iter(sorted(c for c in colors if c.index is None))
+    colors_by_index = {}
+    for c in colors:
+        if c.index is not None:
+            color = c.default()
+            if c.index in colors_by_index and colors_by_index[c.index] != color:
+                raise ValueError(
+                    f"Palette entry index {c.index} already used for "
+                    f"{colors_by_index[c.index]}"
+                )
+            colors_by_index[c.index] = color
+    black = Color(0, 0, 0, 1.0)
+    colors = [
+        (
+            colors_by_index[i]
+            if i in colors_by_index
+            else next(iter_colors_without_index, black)
+        )
+        for i in range(max(colors_by_index, default=-1) + 1)
+    ] + list(iter_colors_without_index)
```

Comment:
I found this a touch hard to follow. Suggested alternate implementation, which may simply prove I didn't follow at all:

```python

def _color_key(color):
  # sort indexed colors by their index, unindexed colors => end
  # all unindexed colors with the same rgb will group together
  if color.has_index:
    return (color.index,)
  return (len(colors), color[:3]) 

palette = []
for key, color in groupby(sorted(colors, key=_color_key), _color_key):
  color = tuple(color)
  if len(color) != 1:
    raise ValueError(f"Multiple colors are assigned {key}: {colors}")
  color = color[0]
  # if there are unused indices blow up
  if len(key) == 1:
    assert len(palette) == key[0], f"Color for index {key[0]} landed at index {len(palette)}"
  palette.append(color)
```

Reply:
The intent of that a bit convoluted code was to keep the color indices of the CPAL we generate as originally specified by the var(--color{index}) variables in the input SVG, but also sort the ones that do not specify an explicit index, placing the in the empty slots in between those with explicit index and any remaining ones appended at the end.

Reply:
Ah, so my first attempt above doesn't do the part about filling in holes. #432 shows my second attempt as I still find the current code a bit hard to follow. Feel free to use or ignore.

Reply:
sorry but I think your code is even more convoluted than mine, perhaps it's a matter of style/opinions..

Reply:
> your code is even more convoluted than mine

I was a little worried that might be the case lol